lib/whitelist/whitelist.rb

```diff
@@ -0,0 +1,43 @@
+class Whitelist
+
+  def initialize(whitelist)
+    @whitelist = whitelist
+  end
+
+  def apply(check_name, headers, rows)
+    check = get_or_else(@whitelist, check_name, {})
+    rules = get_or_else(check, 'rules', [])
+    rule_predicates = rules.map { |rule| Whitelist.predicate_for(headers, get_or_else(rule, 'predicate', [])) }
+    rows.reject { |row| rule_predicates.any? { |r| r.call(row) } }
+  end
+
+  def get_or_else(hash, key, default)
+    # can't use #fetch since it returns nil if the key exists but maps to nil
+    value = hash[key]
+    value ? value : default
+  end
+
+  def report_expired_entries(reference_date)
+    expiries = @whitelist.flat_map { |check| get_or_else(check[1], 'rules', []).map { |rule| [check[0], rule['expiry'], rule['reason']]} }
```

Comment:
I find it hard to scan long lines like this because you have to pick out the method calls before it makes sense.

When I first read this I didn't even realise the second map was part of the first block because I saw some dots and methods and just assumed it was method chaining.

It might help if you extracted a method from this, but I think splitting over multiple lines would make it a lot clearer.

Something like
```ruby
@whitelist.flat_map do |check_name, check|
  check.fetch('rules', []).map do |rule|
    [rule_name, rule['expiry'], rule['reason']]
  end
end